runtime/src/main/java/org/corfudb/runtime/collections/TxnContext.java

```diff
@@ -0,0 +1,387 @@
+    Set<K> keySet(@Nonnull final Table<K, V, M> table) {
+        txBegin(table); // Apply pending writes (if any) & begin corfu txn if thread hasn't begun one
+        table.numKeySets.incrementAndGet();
+        return table.keySet();
+    }
+
+    /**
+     * Scan and filter by entry.
+     *
+     * @param entryPredicate Predicate to filter the entries.
+     * @return Collection of filtered entries.
+     */
+    public <K extends Message, V extends Message, M extends Message>
+    List<CorfuStoreEntry<K, V, M>> executeQuery(@Nonnull final Table<K, V, M> table,
+                                                @Nonnull final Predicate<CorfuStoreEntry<K, V, M>> entryPredicate) {
+        txBegin(table); // Apply pending writes (if any) & begin corfu txn if thread hasn't begun one
+        table.numScans.incrementAndGet();
+        return table.scanAndFilterByEntry(entryPredicate);
+    }
+
+    /**
+     * Start the actual corfu transaction if started by a read operation,
+     * apply all pending operations after starting the corfu transaction.
+     * @param tableBeingRead - only read transactions set this value
+     */
+    private <K extends Message, V extends Message, M extends Message>
+    void txBegin(Table<K, V, M> tableBeingRead) {
+        if (tableBeingRead != null) { // called from one of the read api
+            if (!operations.isEmpty()) { // We have buffered some writes before..
+                operations.forEach(Runnable::run); // apply them for dirty reads to see
+                operations.clear();
+                txnType = TxnType.writeAndReads;
+            }
+            if (txnType == TxnType.writeAndReads) {
+                // Not the first dirty read, don't change the txnType here
+            } else { // Called from read api where no writes have been seen
+                txnType = TxnType.readOnly;
+            }
+            tablesInTxn.add(tableBeingRead);
+        }
+        if (TransactionalContext.isInTransaction()) {
+            if (tableBeingRead == null  // i.e, called from commit() but corfu txn was started!
+                    && !operations.isEmpty()) { // We have pending writes to apply
+                txnType = TxnType.writeAndReads; // Reads, then writes and now commit()
+            }
+            return;
+        }
+        this.txnStartTime = System.nanoTime();
+        Transaction.TransactionBuilder transactionBuilder = this.objectsView
+                .TXBuild()
+                .type(TransactionType.WRITE_AFTER_WRITE);
+        if (isolationLevel.getTimestamp() != Token.UNINITIALIZED) {
+            transactionBuilder.snapshot(isolationLevel.getTimestamp());
+        }
+        transactionBuilder.build().begin();
+    }
+
+    /**
+     * Explicitly end a read only transaction to clean up resources
+     * Internally used at the end of commit() to commit the corfu transaction
+     * @return the address at which the corfu transaction committed.
+     */
+    public long txEnd() {
```

Comment:
can this be made private?

Reply:
Since we agreed that read-your-writes will be the default, we need a public method for query transactions to end()